[CI] IndexFoldersDeletionListenerIT testListenersInvokedWhenIndexIsDangling failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3520 / ubuntu-1804_platform-support-unix](https://gradle-enterprise.elastic.co/s/bwgdzyfvpy2ra)
- [elasticsearch-periodic-platform-support #3496 / windows-2022_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/l5uljyqe3sd7m)
- [elasticsearch-periodic-platform-support #3452 / ubuntu-1804_platform-support-unix](https://gradle-enterprise.elastic.co/s/gxqgo3eia2tys)

**Reproduction Line:**
```
./gradlew ":server:internalClusterTest" --tests "org.elasticsearch.plugins.IndexFoldersDeletionListenerIT.testListenersInvokedWhenIndexIsDangling" -Dtests.seed=9E114B7E93BCFA4C -Dtests.locale=hi -Dtests.timezone=America/Halifax -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testListenersInvokedWhenIndexIsDangling'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.plugins.IndexFoldersDeletionListenerIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: java.lang.AssertionError: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 0
```

**Issue Reasons:**
- [main] 3 failures in test testListenersInvokedWhenIndexIsDangling (0.4% fail rate in 692 executions)
- [main] 2 failures in step ubuntu-1804_platform-support-unix (11.1% fail rate in 18 executions)
- [main] 3 failures in pipeline elasticsearch-periodic-platform-support (16.7% fail rate in 18 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/server/src/internalClusterTest/java/org/elasticsearch/plugins/IndexFoldersDeletionListenerIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/plugins/IndexFoldersDeletionListenerIT.java
@@ -104,7 +104,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
                     plugin.deletedIndices.contains(index)
                 );
 
-                final List<ShardId> deletedShards = plugin.deletedShards.get(index);
+                final Set<ShardId> deletedShards = plugin.deletedShards.get(index);
                 assertThat(deletedShards, notNullValue());
                 assertFalse(
                     "Listener should have been notified of deletion of one or more shards on node " + nodeName,
@@ -175,7 +175,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
                         plugin.deletedIndices.contains(index)
                     );
 
-                    final List<ShardId> deletedShards = plugin.deletedShards.get(index);
+                    final Set<ShardId> deletedShards = plugin.deletedShards.get(index);
                     assertThat(deletedShards, notNullValue());
                     assertFalse(
                         "Listener should have been notified of deletion of one or more shards on node " + nodeName,
@@ -338,7 +338,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
     public static class IndexFoldersDeletionListenerPlugin extends Plugin implements IndexStorePlugin {
 
         final Set<Index> deletedIndices = ConcurrentCollections.newConcurrentSet();
-        final Map<Index, List<ShardId>> deletedShards = ConcurrentCollections.newConcurrentMap();
+        final Map<Index, Set<ShardId>> deletedShards = ConcurrentCollections.newConcurrentMap();
 
         @Override
         public List<IndexFoldersDeletionListener> getIndexFoldersDeletionListeners() {
@@ -350,7 +350,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
 
                 @Override
                 public void beforeShardFoldersDeleted(ShardId shardId, IndexSettings indexSettings, Path[] shardPaths) {
-                    deletedShards.computeIfAbsent(shardId.getIndex(), i -> new ArrayList<>()).add(shardId);
+                    deletedShards.computeIfAbsent(shardId.getIndex(), i -> ConcurrentCollections.newConcurrentSet()).add(shardId);
                 }
             });
         }

--- a/server/src/internalClusterTest/java/org/elasticsearch/plugins/IndexFoldersDeletionListenerIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/plugins/IndexFoldersDeletionListenerIT.java
@@ -104,7 +104,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
                     plugin.deletedIndices.contains(index)
                 );
 
-                final Set<ShardId> deletedShards = plugin.deletedShards.get(index);
+                final List<ShardId> deletedShards = plugin.deletedShards.get(index);
                 assertThat(deletedShards, notNullValue());
                 assertFalse(
                     "Listener should have been notified of deletion of one or more shards on node " + nodeName,
@@ -175,7 +175,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
                         plugin.deletedIndices.contains(index)
                     );
 
-                    final Set<ShardId> deletedShards = plugin.deletedShards.get(index);
+                    final List<ShardId> deletedShards = plugin.deletedShards.get(index);
                     assertThat(deletedShards, notNullValue());
                     assertFalse(
                         "Listener should have been notified of deletion of one or more shards on node " + nodeName,
@@ -338,7 +338,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
     public static class IndexFoldersDeletionListenerPlugin extends Plugin implements IndexStorePlugin {
 
         final Set<Index> deletedIndices = ConcurrentCollections.newConcurrentSet();
-        final Map<Index, Set<ShardId>> deletedShards = ConcurrentCollections.newConcurrentMap();
+        final Map<Index, List<ShardId>> deletedShards = ConcurrentCollections.newConcurrentMap();
 
         @Override
         public List<IndexFoldersDeletionListener> getIndexFoldersDeletionListeners() {
@@ -350,7 +350,7 @@ public class IndexFoldersDeletionListenerIT extends ESIntegTestCase {
 
                 @Override
                 public void beforeShardFoldersDeleted(ShardId shardId, IndexSettings indexSettings, Path[] shardPaths) {
-                    deletedShards.computeIfAbsent(shardId.getIndex(), i -> ConcurrentCollections.newConcurrentSet()).add(shardId);
+                    deletedShards.computeIfAbsent(shardId.getIndex(), i -> Collections.synchronizedList(new ArrayList<>())).add(shardId);
                 }
             });
         }

